### PR TITLE
Erowlin/fix default config values and do not trigger double sql queries when refresh token is disabled

### PR DESCRIPTION
**Pull request by Erowlin**
### Summary

This solves #1099 : Whenever authenticating on the server, we tried to revoke a refresh token, even if it was not enabled in the configuration (initializer). 
It removes the unnecessary call to the ORM (or the storage). 

I also updated the `config` builder because it were returning tri-state booleans (`nil`, `true`, `false`) instead of `true` or `false`. Methods finishing with `question mark ?` should never return nil. 

### Other Information

Should I update `NEWS.md`?

### CODE REVIEW

File: lib/doorkeeper/config.rb [hunk 1]
```diff
@@ -292,24 +292,32 @@ def option(name, options = {})
     attr_reader :api_only
     attr_reader :enforce_content_type
 
+    def api_only
+      @api_only ||= false
+    end
+
+    def enforce_content_type
```
[the remark this answers is not in the document]
Comment (Erowlin) [reply lib/doorkeeper/config.rb]: This one is fine, it is to set a default value to the `enforce_content_type`. I t is conflicting with the attr_reader. 
How should I fix that / ignore the message?

File: lib/doorkeeper/config.rb [hunk 2]
```diff
@@ -292,24 +292,32 @@ def option(name, options = {})
     attr_reader :api_only
     attr_reader :enforce_content_type
 
+    def api_only
```
[the remark this answers is not in the document]
Comment (Erowlin) [reply lib/doorkeeper/config.rb]: This one is fine, it is to set a default value to the enforce_content_type. I t is conflicting with the attr_reader.
How should I fix that / ignore the message?

**Comment by nbulaj** [comment]
Thanks for contribution! I will check it in a hour.

> Should I update NEWS.md?

Sure! And then please squash all the commits to a single one. Thanks!

**Comment by Erowlin** [comment]
@nbulaj It's done. Let me know what's next :).

File: NEWS.md [hunk 3]
```diff
@@ -6,7 +6,10 @@ upgrade guides.
 User-visible changes worth mentioning.
 
 ## master
-
+- [#1099] All the configuration variables in `Doorkeeper.configuration` now always return
```
Comment (nbulaj) [inline NEWS.md]: `always return` duplicates :)

File: lib/doorkeeper/config.rb [hunk 4]
```diff
@@ -292,24 +292,32 @@ def option(name, options = {})
...
+      @api_only ||= false
+    end
+
+    def enforce_content_type
+      @enforce_content_type ||= false
+    end
+
     def refresh_token_enabled?
-      defined?(@refresh_token_enabled) && @refresh_token_enabled
+      !!(defined?(@refresh_token_enabled) && @refresh_token_enabled)
```
Comment (nbulaj) [inline lib/doorkeeper/config.rb]: I think in this case we can return to previous  `!!@refresh_token_enabled` solution, but what is the case? `nil`-values in ruby is always falsey, what is the purpose of this changes?

File: lib/doorkeeper/oauth/token.rb [hunk 5]
```diff
@@ -4,16 +4,19 @@ class Token
       class << self
         def from_request(request, *methods)
           methods.inject(nil) do |_, method|
-            method = self.method(method) if method.is_a?(Symbol)
+            method      = self.method(method) if method.is_a?(Symbol)
```
Comment (nbulaj) [inline lib/doorkeeper/oauth/token.rb]: No need in this indent

File: lib/doorkeeper/oauth/token.rb [hunk 6]
```diff
@@ -4,16 +4,19 @@ class Token
...
             credentials = method.call(request)
             break credentials unless credentials.blank?
           end
         end
 
         def authenticate(request, *methods)
           if (token = from_request(request, *methods))
-            access_token = AccessToken.by_token(token)
-            access_token.revoke_previous_refresh_token! if access_token
+            access_token          = AccessToken.by_token(token)
```
Comment (nbulaj) [inline lib/doorkeeper/oauth/token.rb]: Also no need in indenttion

File: lib/doorkeeper/config.rb (hunk 4 above)
Comment (Erowlin) [reply lib/doorkeeper/config.rb]: The point is to avoid tri-state booleans as described here: https://robots.thoughtbot.com/avoid-the-threestate-boolean-problem. (The article is about DB, but I apply the same rule for Ruby code. 
Also as a rule of thumb, all question-marks methods should return either `true` or `false`. Very opinionated though, I can revert it if you're not comfortable :).

File: lib/doorkeeper/oauth/token.rb (hunk 5 above)
Comment (Erowlin) [reply lib/doorkeeper/oauth/token.rb]: This has been reverted already.

File: lib/doorkeeper/oauth/token.rb (hunk 6 above)
Comment (Erowlin) [reply lib/doorkeeper/oauth/token.rb]: Reverted as well.

File: lib/doorkeeper/config.rb (hunk 4 above)
Comment (nbulaj) [reply lib/doorkeeper/config.rb]: No no, I think it's a reasonable one. So maybe we can just return `!!@var` syntax for the options?

Comment (Erowlin) [reply lib/doorkeeper/config.rb]: I'm not sure I get what you are suggesting. 

Can you provide an example?

Comment (nbulaj) [reply lib/doorkeeper/config.rb]: Previously it was just

```ruby
!!@refresh_token_enabled
```

instead of

```ruby
defined?(@refresh_token_enabled) && @refresh_token_enabled
```

Comment (nbulaj) [reply lib/doorkeeper/config.rb]: A little bit another way :)

https://github.com/doorkeeper-gem/doorkeeper/blob/ed95d941e5846ed5b8ef90ff2f6bac96f83e2fec/lib/doorkeeper/config.rb#L284-L285

So maybe we can save the code from your commit

Comment (Erowlin) [reply lib/doorkeeper/config.rb]: Ok I get what you meant.

Comment (Erowlin) [reply lib/doorkeeper/config.rb]: Ok I get what you meant. The issue with this approach is that the Doorkeeper.configuration object returned will contains the wrong value, it can be misleading to have the getter methods returning false and the Configuration object having `nil`.

File: lib/doorkeeper/config.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (nbulaj) [reply lib/doorkeeper/config.rb]: So maybe we need to change `attr_reader` to `attr_writer` because we already have a reader method?

File: lib/doorkeeper/config.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbulaj) [reply lib/doorkeeper/config.rb]: The same

File: lib/doorkeeper/config.rb (hunk 4 above)
Comment (nbulaj) [reply lib/doorkeeper/config.rb]: I think we can save you changes

Comment (Erowlin) [reply lib/doorkeeper/config.rb]: So, do I have something to change or not? 
Sorry, I'm super confused on what you need me to do, and it's also the end of the day in France ;)

**Comment by Erowlin** [comment]
I squashed the commits, I think it's fine now.

**Comment by Erowlin** [comment]
Quick question, if I update my gem on my project, does it contains this fix yet?

**Comment by nbulaj** [comment]
Only if it is set up from git. Stable release on RubyGems will be later.

**Comment by Erowlin** [comment]
Ok, thank you!
